Proposal:  Extend .option directive for control enabled extensions on specific code region (#67)
riscv-non-isa/riscv-asm-manual: riscv-asm.md

.p2align     | p2,[pad_val=0],max             | align to power of 2
.balign      | b,[pad_val=0]                  | byte align
.zero        | integer                        | zero bytes
.push_arch   | extension_list                 | Push current extension status to stack, and enable all extension listed in the extension list, detail syntax see [`.push_arch`/`.pop_arch`](<#push-pop-arch).
.pop_arch    |                                | Pop current extension status from stack.

### <a name=push-pop-arch></a>`.push_arch`/`.pop_arch`

`.push_arch` / `.pop_arch` are used to control enabled ISA extension for certain
code regions, but without change the arch attribute, that means it won't raise
the minimal execution environment requirement, so user should take care to
execution the code regions around `.push_arch`/`.pop_arch`, typical use case is
used with `ifunc`, e.g. libc are built with `rv64gc`, but few functions like
memcpy provide two version, one built with `rv64gc`, and one built with
`rv64gcv`, and select by ifunc mechanism at run-time, however we don't want to

> change to minimal execution environment requirement to `rv64gcv`, since

Reviewer: ```suggestion
change the minimal execution environment requirement to `rv64gcv`, since
```

> `rv64gcv` version will invoked only if execution environment is supported

Reviewer: ```suggestion
the `rv64gcv` version will be invoked only if the execution environment supports
```

> for vector extension, so the minimal execution environment requirement still is

Reviewer: ```suggestion
the vector extension, so the minimal execution environment requirement still is
```

`rv64gc`.

```

> EXTENSION_LIST := EXTENSION_LIST ',' EXTENSION_LIST

Reviewer: ```suggestion
EXTENSION_LIST := EXTENSION_LIST ',' EXTENSION
```

.balign      | b,[pad_val=0]                  | byte align
.zero        | integer                        | zero bytes

### <a name=.option></a> `.option`

#### `rvc`/`norvc`

Enable/disable C-extension for following code region, this option is equivalent
to `.option arch +c`/`.option arch -c`, but widely support by older version of
toolchain.

#### `arch`

Enable and/or disable specific ISA extensions for following code regions, but
without changing the arch attribute, that means it won't raise the minimal
execution environment requirement, so the user should take care to protect
execution of the code regions around `.option push`/`.option arch`/`.option pop`.
A typical use case is is with `ifunc`, e.g. the libc is built with `rv64gc`,
but a few functions like memcpy provide two versions, one built with `rv64gc`
and one built with `rv64gcv`, and then select between them by ifunc mechanism
at run-time.  However we don't want to change the minimal execution environment
requirement to `rv64gcv`, since the `rv64gcv` version will be invoked only if
the execution environment supports the vector extension, so the minimal

> execution environment requirement still is `rv64gc`.

Reviewer 2: This should be commentary.

Reviewer 3: If this is changed to commentary, it's important that the part about "won't raise the minimal execution environment requirement" be left as normative. We don't want to set `EF_RISCV_RVC` in the ELF flags when using `.option arch, +c`, even though binutils currently does that for `.option rvc`. See https://reviews.llvm.org/D122490#3419774 for a recent discussion about that.

.zero        | integer                        | zero bytes
.variant_cc  | symbol_name                    | annotate the symbol with variant calling convention

### <a name=.option></a> `.option`

#### `rvc`/`norvc`

This option will be deprecated soon once `.option arch` is widely implemented on
main stream open source toolchains.

Enable/disable the C-extension for the following code region, this option is equivalent
to `.option arch, +c`/`.option arch, -c`, but widely support by older version of
toolchain.

NOTE: `.option rvc` might set the `EF_RISCV_RVC` in the ELF flags on specific
toolchains, that might cause linker convert instruction into compressed
instruction on unexpected region.

NOTE: The difference between `.option rvc`/`.option norvc` and

> `.option arch +c`/`.option arch -c` won't set `EF_RISCV_RVC` in the ELF flags.

Reviewer 3: NOTE: There is a difference between `.option rvc`/`.option norvc` and
`.option arch +c`/`.option arch -c`. The latter won't set `EF_RISCV_RVC` in the ELF flags.

#### `arch`

> Enable and/or disable specific ISA extensions for following code regions, but

Reviewer 3: "for following" -> "for the following"

invoked only if the execution environment supports the vector extension, so
the minimal execution environment requirement still is `rv64gc`.

NOTE: `.option arch, +` will also enable all required extensions, for example,
`rv32i` + `.option arch, +v` will also enable `f`, `d`, `zve32x`, `zve32f`,
`zve64x`, `zve64f`, `zve64d`, `zvl32b`, `zvl64b` and `zvl128b` extensions.

NOTE: We recommended `.option arch, +` and `.option arch, -` are used with
`.option push`/`.option pop` instead of a `.option arch, +` / `.option arch, -`
pair, because `.option arch, +` will enable all required extensions, but
`.option arch, -` only disable specific extension, so the result might be
unexpected, for example: `rv32i` + `.option arch, +v` + `.option arch, -v`
will result `rv32ifd_zve32x_zve32f_zve64x_zve64f_zve64d_zvl32b_zvl64b_zvl128b`
not `rv32i`.

#### `pic`/`nopic`

Set the code model to PIC (position independent code) or non-PIC. This will
affect the expansion of the `la` pseudoinstruction, refer to
[listing of standard RISC-V pseudoinstructions](#pseudoinstructions).

#### `relax`/`norelax`

Enable/disable linker relaxation for the following code region.

> NOTE: Code region follows by `.option relax` will emit

Reviewer 3: "follows by" -> "followed by"

> `R_RISCV_RELAX`/`R_RISCV_ALIGN` even linker unsupport relaxation, suggested

Reviewer 3: "even linker unsupport relaxation" -> "even if the linker does not support relaxation"?

#### `rvc`/`norvc`

This option will be deprecated soon after `.option arch` has been widely
implemented on main stream open source toolchains.

Enable/disable the C-extension for the following code region. This option is
equivalent to `.option arch, +c`/`.option arch, -c`, but widely support by
older toolchain versions.

Alternative style:
NOTE: .option rvc might set the ELF flag EF_RISCV_RVC in some toolchains. That
might cause the linker to compress instructions in code regions where that was
not intended.

NOTE: There is a difference between `.option rvc`/`.option norvc` and
`.option arch +c`/`.option arch -c`. The latter won't set EF_RISCV_RVC in the

> ELF flags.

Author: `EF_RISCV_RVC` should be deprecated IMO, linker currently rely on that to decide linker relaxation can use C instruction or not, but that apparently not a completeness solution.

Give an example to show what's wrong:
```
.attribute arch, "rv64g" # NO RVC!

...
# Only allow rvc in following region
.option rvc
...
.option norvc
...
```

However option rvc will set `EF_RISCV_RVC` will cause whole program might relaxed into C instruction even not in the rvc/norvc region.

> ELF flags.

Reviewer 4: Agreed.  FYI, there are two patches might help to clarify the idea,

* RISC-V: File-level architecture shouldn't be affected by section-level ones
https://sourceware.org/pipermail/binutils/2022-November/124178.html

* RISC-V: Refer mapping symbol to R_RISCV_RELAX for rvc relaxations
https://sourceware.org/pipermail/binutils/2022-September/123191.html

#### `arch`

Enable and/or disable specific ISA extensions for the following code regions, but
without changing the arch attribute and `EF_RISCV_RVC` in the ELF flags, that
means it won't raise the minimal execution environment requirement, so the user
should take care to protect execution of the code regions around
`.option push`/`.option arch`/`.option pop`.

Syntax for `.option arch`:
```

> .option arch, <EXTENSION>

Reviewer 5: binutils has:
```
.option arch, <EXTENSION-LIST>

EXTENSION-LIST := <EXTENSION>
                | <EXTENSION>, <EXTENSION-LIST>
```

OP             := '+'
                | '-'

EXTENSION      := OP <extension-name> VERSION
                | OP <extension-name>

>                 | '=' <full-arch-string>

Reviewer 5: binutils dropped the leading =

`rv64gc`, but a few functions like memcpy provide two versions, one built with
`rv64gc` and one built with `rv64gcv`, and then select between them by ifunc
mechanism at run-time.  However we don't want to change the minimal execution
environment requirement to `rv64gcv`, since the `rv64gcv` version will be
invoked only if the execution environment supports the vector extension, so
the minimal execution environment requirement still is `rv64gc`.

NOTE: `.option arch, +` will also enable all required extensions, for example,
`rv32i` + `.option arch, +v` will also enable `f`, `d`, `zve32x`, `zve32f`,
`zve64x`, `zve64f`, `zve64d`, `zvl32b`, `zvl64b` and `zvl128b` extensions.

NOTE: We recommended `.option arch, +` and `.option arch, -` are used with
`.option push`/`.option pop` instead of a `.option arch, +` / `.option arch, -`
pair, because `.option arch, +` will enable all required extensions, but
`.option arch, -` only disable specific extension, so the result might be
unexpected, for example: `rv32i` + `.option arch, +v` + `.option arch, -v`
will result `rv32ifd_zve32x_zve32f_zve64x_zve64f_zve64d_zvl32b_zvl64b_zvl128b`

> not `rv32i`.

Reviewer 5: More clarity is needed for:
```
.option arch, rv64ifd
.option arch, -f
```
Currently binutils will treat that that as a no-op, since it will turn it into rv64id and then immediately re-compute the implications. I view this as surprising and not that helpful; ideally it would result in rv64i instead. Otherwise you need to know the full list of possible extensions and list all the ones _that the current toolchain supports_, which will change over time as new extensions get added (like how parts of V and Zfh were added after F).

Author: Add example for this case.

#### `rvc`/`norvc`

This option will be deprecated soon after `.option arch` has been widely
implemented on main stream open source toolchains.

Enable/disable the C-extension for the following code region. This option is
equivalent to `.option arch, +c`/`.option arch, -c`, but widely support by
older toolchain versions.

Alternative style:

```
.option push

> .option +c   # Alternative of .option rvc

Reviewer 6: `.option arch, +c`

I think you do not intend to allow `.option +c`, right?

Author: Fixed

.option pop

.option push
.option -c   # Alternative of .option norvc
.option pop
```

> NOTE: .option rvc might set the ELF flag EF_RISCV_RVC in some toolchains. That

Reviewer 6: backticks around `.option rvc` and `EF_RISCV_RVC`.

might cause the linker to compress instructions in code regions where that was
not intended.

NOTE: There is a difference between `.option rvc`/`.option norvc` and

> `.option arch +c`/`.option arch -c`. The latter won't set EF_RISCV_RVC in the

Reviewer 6: `.option arch, +c`

Better not to allow `.option .arch +c` (space-separated).

Author: Updated, that's typo, I am not intend to support space-separated form

ELF flags.

#### `arch`

Enable and/or disable specific ISA extensions for the following code regions, but
without changing the arch attribute and `EF_RISCV_RVC` in the ELF flags, that
means it will not raise the minimal execution environment requirement, so the user
should take care to the execution of the code regions around
`.option push`/`.option arch`/`.option pop`.

Syntax for `.option arch`:
```

.option arch, <EXTENSIONS-OR-FULLARCH>

EXTENSIONS-OR-FULLARCH := <EXTENSIONS>
                        | <FULLARCHSTR>

EXTENSIONS             := <EXTENSION> ',' <EXTENSIONS>
                        | <EXTENSION>

FULLARCHSTR            := <full-arch-string>

EXTENSION              := <OP> <EXTENSION-NAME> [<VERSION>]

OP                     := '+'
                        | '-'

VERSION                := [0-9]+ 'p' [0-9]+

>                         | [1-9][0-9]*

Reviewer 3: I'm not sure this is correct. Can you please walk me through it?

Author: I guess the current rule might be confused by `[`, `]`, so I try to rewrite the rule to allow VERSION as empty string, also expand first rule into two for forbid `0p0`.

```
EXTENSION              := <OP> <EXTENSION-NAME> [<VERSION>]

VERSION                := [0-9]+ 'p' [0-9]+
                        | [1-9][0-9]*
```

to

```
EXTENSION              := <OP> <EXTENSION-NAME> <VERSION>

VERSION                := [1-9][0-9]* 'p' [0-9]+
                        | [0-9]+ 'p' [1-9][0-9]*
                        | [1-9][0-9]*
                        |
```

Author: @luismarques let me know it's still unclear to you or any case you think it's still ambiguous

Reviewer 3: LGTM now.

EXTENSION-NAME         := Naming rule is defined in RISC-V ISA manual
```

- Extension version can be omitted, the assembler will use the built-in default
  version for that extension.
- `OP` can be enable (`+`) or disable (`-`).

> - Input format of `<full-arch-string>` is the same as `-march` option.

Reviewer 6: The format of

I think `input` can be omitted.

Example:

```assembly

> .attribute arch, "rv64imafdc"

Reviewer 3: Update to match binutils, remove quotes.

> # You can only use instruction from i, m, a, f, d and c extensions.

Reviewer 6: instructions from the

memcpy_general:
    add     a5,a1,a2
    beq     a1,a5,.L2
    add     a2,a0,a2
    mv      a5,a0
.L3:
    addi    a1,a1,1
    addi    a5,a5,1
    lbu     a4,-1(a1)
    sb      a4,-1(a5)
    bne     a5,a2,.L3
.L2:
    ret

> .option push     # Push current option to stack.

Reviewer 6: Push current options to the stack.

.option arch, +v # Enable vector extension, we can use any instruction in imafdcv extension.
memcpy_vec:
    mv a3, a0
.Lloop:
    vsetvli t0, a2, e8, m8, ta, ma
    vle8.v v0, (a1)
    add a1, a1, t0
    sub a2, a2, t0
    vse8.v v0, (a3)
    add a3, a3, t0
    bnez a2, .Lloop
    ret
.option pop   # Pop current option from stack, restore the enabled ISA extension status to imafdc.

.option push     # Push current option to stack.
.option arch, -c # Disable compressed extension, we can't use any instruction in extension.
memcpy_norvc:
    add     a5,a1,a2
    beq     a1,a5,.L2
    add     a2,a0,a2
    mv      a5,a0
.L3:
    addi    a1,a1,1
    addi    a5,a5,1
    lbu     a4,-1(a1)
    sb      a4,-1(a5)
    bne     a5,a2,.L3
.L2:
    ret
.option pop   # Pop current option from stack, restore the enabled ISA extension status to imafdc.

> .option push           # Push current option to stack.

Reviewer 6: misaligned `#`

.option arch, rv64imc # Set arch to rv64imc.
    nop
.option pop   # Pop current option from stack, restore the enabled ISA extension status to imafdc.

```

NOTE: A typical use case is with `ifunc`, e.g. the C library is built with
`rv64gc`, but a few functions like memcpy provide two versions, one built with
`rv64gc` and one built with `rv64gcv`, and then select between them by ifunc
mechanism at run-time.  However, we don't want to change the minimal execution
environment requirement to `rv64gcv`, since the `rv64gcv` version will be
invoked only if the execution environment supports the vector extension, so
the minimal execution environment requirement still is `rv64gc`.

NOTE: `.option arch, +` will also enable all required extensions, for example,
`rv32i` + `.option arch, +v` will also enable `f`, `d`, `zve32x`, `zve32f`,

> `zve64x`, `zve64f`, `zve64d`, `zvl32b`, `zvl64b` and `zvl128b` extensions.

Reviewer 3: It would be good to explain somewhere that e.g. `+c, -c` and `+v, -v` do the expected things. I guess the expected thing is that the named extension is disabled but its dependencies are enabled.

Reviewer 3: (alternatively it could be an error)

> NOTE: We recommended `.option arch, +` and `.option arch, -` are used with

Reviewer 3: ```suggestion
NOTE: We recommend `.option arch, +` and `.option arch, -` are used with
```

same result `rv64ifd`, because `f` will add back during compute implication
extensions for `rv64id`.

> NOTE: `.option arch, +<ext>, -<ext>` can be used to enable dependency
> extensions, e.g. `rv32i` + `.option arch, +v, -v` will result
> `rv32ifd_zve32x_zve32f_zve64x_zve64f_zve64d_zvl32b_zvl64b_zvl128b`.

Reviewer 3: ```suggestion
NOTE: `.option arch, +<ext>, -<ext>` is accepted and will result in enabling the
extensions that depend on `ext`, e.g. `rv32i` + `.option arch, +v, -v` will result
`rv32ifd_zve32x_zve32f_zve64x_zve64f_zve64d_zvl32b_zvl64b_zvl128b`.
```
